src/cpp/src/model_runner.hpp

```diff
@@ -92,33 +111,35 @@ class ModelRunner {
                     position_ids_data[token_id] = position_id;
                 }
 
-                past_lens_data[0] = group_context_len;
+                size_t expected_kv_cache_size = sequence_group->get_num_processed_tokens() - sequence->get_num_evicted_tokens();
+                past_lens_data[0] = expected_kv_cache_size;
 
                 subsequence_begins_data[1] = subsequence_begins_data[0] + num_scheduled_tokens;
-                block_indices_begins_data[1] = block_indices_begins_data[0] + num_blocks;
 
-                const std::vector<KVCacheBlock::Ptr> & kv_blocks = scheduler_output.m_block_tables.at(sequence->get_id());
-                for (size_t block_id = 0; block_id < num_blocks; ++block_id)
-                    block_indices_data[block_id] = kv_blocks[block_id]->get_index();
+                size_t num_blocks = (sequence_group->get_context_len()  - sequence->get_num_evicted_tokens() +  m_scheduler_config.block_size - 1) / m_scheduler_config.block_size;
+                block_indices_begins_data[1] = block_indices_begins_data[0] + num_blocks;
 
                 // apply strides to shift to a next sequence
                 input_ids_data += num_scheduled_tokens;
                 position_ids_data += num_scheduled_tokens;
                 past_lens_data += 1;
                 subsequence_begins_data += 1;
-                block_indices_data += num_blocks;
                 block_indices_begins_data += 1;
             }
         }
 
+
+
         // typical LLM parameters
         m_request.set_tensor("input_ids", input_ids);
         m_request.set_tensor("position_ids", position_ids);
 
         // PA specific parameters
         m_request.set_tensor("past_lens", past_lens);
         m_request.set_tensor("subsequence_begins", subsequence_begins);
-        m_request.set_tensor("block_indices", block_indices);
+
+        m_request = _set_block_indices(m_request, sequence_groups, scheduler_output, total_num_blocks);
```

Comment:
we don't need to pass / return m_request here, because it's a class field

Reply:
Done